mmv1/third_party/terraform/transport/transport.go

```diff
@@ -189,3 +189,130 @@ func IsApiNotEnabledError(err error) bool {
 	}
 	return false
 }
+
+func PluralDataSourceGetList(d *schema.ResourceData, config *Config, billingProject *string, userAgent string, u string, listFlattener func(config *Config, res interface{}) ([]interface{}, error), params map[string]string, resourecToList string) ([]interface{}, error) {
+	if params == nil {
+		params = make(map[string]string)
+	}
+
+	items := make([]interface{}, 0)
+	for {
+		// Depending on previous iterations, params might contain a pageToken param
+		url, err := AddQueryParams(u, params)
+		if err != nil {
+			return nil, err
+		}
+
+		headers := make(http.Header)
+		opts := SendRequestOptions{
+			Config:               config,
+			Method:               "GET",
+			RawURL:               url,
+			UserAgent:            userAgent,
+			Headers:              headers,
+			ErrorRetryPredicates: []RetryErrorPredicateFunc{Is429RetryableQuotaError},
+		}
+		if billingProject != nil {
+			opts.Project = *billingProject
+		}
+		res, err := SendRequest(opts)
+		if err != nil {
+			return nil, HandleNotFoundError(err, d, fmt.Sprintf("%s %q", resourecToList, d.Id()))
+		}
+
+		if res == nil {
```

Comment:
A plural datasource can return zero items and still be valid, I think? I'd expect us to drop the id if the _parent_ is gone i.e. a crypto key with no keys in it still exists, it just has no keys. (for most resources that will be the project or the project+location pair)

Reply:
I would further consider making this the responsibility of the caller, where this function's responsibility is just to walk the paginated list, leaning on my comment about `GetPaginatedList` a bit more!

Reply:
yeah we can change this to where it traverses the list, regardless of whether it contains anything or not.

> I'd expect us to drop the id if the parent is gone i.e. a crypto key with no keys in it still exists, it just has no keys. (for most resources that will be the project or the project+location pair)

this makes sense to check whether the actual parent (resource we need that contains the list) is present.

we completely remove the id of the plural data sources of crypto keys if the `crypto key` itself that contains the list is nonexistent.